src/libPMacc/include/particles/ParticlesBase.kernel

```diff
@@ -350,75 +350,176 @@ struct KernelShiftParticles
 
         if ( lastFrame.isValid( ) )
         {
+            using ParticleDomCfg = IdxConfig<
+                frameSize,
+                numWorkers
+            >;
+
+            /* context if a element within the frame is a particle */
+            memory::CtxArray<
+                bool,
+                ParticleDomCfg
+            >
+            isParticleCtx(
+                workerIdx,
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    return lastFrame[ linearIdx ][ multiMask_ ];
+                }
+            );
+
+            /* loop over all particles in the frame */
+            ForEachIdx< ParticleDomCfg > forEachParticle( workerIdx );
+
             //count particles in last frame
-            const bool isParticle = lastFrame[threadIdx.x][multiMask_];
-            if ( isParticle == true ) // \todo: count bits
-            {
-                nvidia::atomicAllInc( &counterParticles );
-            }
+            forEachParticle(
+                [&](
+                    uint32_t const,
+                    uint32_t const idx
+                )
+                {
+                    if( isParticleCtx[ idx ] )
+                        nvidia::atomicAllInc( &counterParticles );
+                }
+            );
+
             __syncthreads( );
 
-            if ( threadIdx.x < counterParticles && isParticle == false )
-            {
-                const int localGapIdx = nvidia::atomicAllInc( &counterGaps );
-                gapIndices_sh[localGapIdx] = threadIdx.x;
-            }
+            forEachParticle(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const idx
+                )
+                {
+                    if ( linearIdx < counterParticles && isParticleCtx[ idx ] == false )
+                    {
+                        int const localGapIdx = nvidia::atomicAllInc( &counterGaps );
+                        gapIndices_sh[ localGapIdx ] = linearIdx;
+                    }
+                }
+            );
             __syncthreads( );
-            if ( threadIdx.x >= counterParticles && isParticle )
+            forEachParticle(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const idx
+                )
+                {
+                    if ( linearIdx >= counterParticles && isParticleCtx[ idx ] )
+                    {
+                        //any particle search a gap
+                        int const srcGapIdx = nvidia::atomicAllInc( &srcGap );
+                        int const gapIdx = gapIndices_sh[ srcGapIdx ];
+                        auto parDestFull = lastFrame[ gapIdx ];
+                        /*enable particle*/
+                        parDestFull[ multiMask_ ] = 1;
+                        /* we not update multiMask because copy from mem to mem is to slow
```

Comment:
"we not" => "we do not"
"copy" => "copying" (not sure, maybe ask someone around you)
"to slow" => "too slow"